Secondary index of collection columns

nyh opened the pull request:
This pull request introduces global secondary-indexing for non-frozen collections.

The intent is to enable such queries:

```
CREATE TABLE test(int id, somemap map<int, int>, somelist<int>, someset<int>, PRIMARY KEY(id));
CREATE INDEX ON test(keys(somemap));
CREATE INDEX ON test(values(somemap));
CREATE INDEX ON test(entries(somemap));
CREATE INDEX ON test(values(somelist));
CREATE INDEX ON test(values(someset));

-- index on test(c) is the same as index on (values(c))
CREATE INDEX IF NOT EXISTS ON test(somelist);    
CREATE INDEX IF NOT EXISTS ON test(someset);
CREATE INDEX IF NOT EXISTS ON test(somemap);

SELECT * FROM test WHERE someset CONTAINS 7;
SELECT * FROM test WHERE somelist CONTAINS 7;
SELECT * FROM test WHERE somemap CONTAINS KEY 7;
SELECT * FROM test WHERE somemap CONTAINS 7;
SELECT * FROM test WHERE somemap[7] = 7;
```

We use here all-familiar materialized views (MVs). Scylla treats all the collections the same way - they're a list of pairs (key, value). In case of sets, the value type is dummy one. In case of lists, the key type is TIMEUUID. When describing the design, I will forget that there is more than one collection type.
Suppose that the columns in the base table were as follows:

```
pkey int, ckey1 int, ckey2 int, somemap map<int, text>, PRIMARY KEY(pkey, ckey1, ckey2)
```

The MV schema is as follows (the names of columns which are not the same as in base might be different). All the columns here form the primary key.

```
-- for index over entries
indexed_coll (int, text), idx_token long, pkey int, ckey1 int, ckey2 int
-- for index over keys
indexed_coll int, idx_token long, pkey int, ckey1 int, ckey2 int
-- for index over values
indexed_coll text, idx_token long, pkey int, ckey1 int, ckey2 int, coll_keys_for_values_index int
```

The reason for the last additional column is that the values from a collection might not be unique.

Fixes #2962
Fixes #8745
Fixes #10707

This patch does not implement **local** secondary indexes for collection columns: Refs #10713.

nyh commented:
This is a clone of @enedil 's work from https://github.com/scylladb/scylla/pull/10123 with a couple of textual and semantic conflicts fixed, and I plan to push it now until it's finally committed ASAP.

Please note that some important review comments are left on pull request #10123 but unfortunately it's not possible (as far as I know) to change the ownership or branch of a PR, so I had to create a new one, so we'll need to figure out what exactly remains to be done here.

nyh commented:
Added a 24th patch to un-xfailed several of the tests I wrote in test_secondary_index.py which start to pass with this series (which is good news).

Only three collection indexing tests continue to xfail:
* `test_secondary_index.py::test_index_collection_wrong_type` (I don't know why)
* `test_secondary_index.py::test_index_quoted_names` (this is #10707)
* `test_secondary_index.py::test_local_secondary_index_on_collection` (this is #10713)

nyh commented:
`test_secondary_index.py::test_index_collection_wrong_type` fails just because the name of the index created by `'CREATE INDEX ON tbl(x)` when x is a collection isn't tbl_x_idx as it is in Cassandra but something else (probably with "values"?). 

I'll create a new test for this incompatibility, and we need to fix it now - before committing this series and people start to have wrong names for the indexes. As I previously explained in #3403, I don't want to have such an incompatibility between Cassandra's and Scylla's default index names, because it can cause scripts which assume they can drop an index by its default name to break.

nyh commented:
Added a 25th patch which reproduces in a test, and fixes, the problem of a wrong index name for collection indexes - the default index name for "CREATE INDEX ON tbl(keys(m))" should be `tbl_m_idx`, not `tbl_m_keys_idx`. 
We had a fairly elaborate function to calculate this "m_keys" string, and this function was simply not needed so this new patch removes it (I could conflate it with @enedil 's original patch, but rather not, it will be easier to review just my patches at the end of the patch list).

nyh commented:
Added 26th patch to finally fix `test_secondary_index.py::test_index_collection_wrong_type`. After the 25th patch fixed the default index naming (which this test assumed), the remaining problem was in the expected error messages - and this patch fixes it.

Interestingly, the fix resulted in removing an abstraction that was called "index_option" that was not necessary at all - and just added to the complexity of the collection indexing code. Dozens of lines across multiple source lines were replaced by a 9-line static function, that also prints strings which are both more useful to users, and match what Cassandra does :-)

After this patch, only two xfailing tests remain - both of them known and can potentially be fixed after committing this PR (because they already have separate issue numbers):

* `test_secondary_index.py::test_index_quoted_names` (https://github.com/scylladb/scylla/issues/10707)
* `test_secondary_index.py::test_local_secondary_index_on_collection` (https://github.com/scylladb/scylla/issues/10713)

psarna commented:
the last two changes look good to me; also, adding @avikivity to the reviewer set, copied from #10123

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/954/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/955/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/958/

nyh commented:
Rebased and pushed a 27th patch fixing issue https://github.com/scylladb/scylla/issues/10707 (indexing of columns with weird names).

The patch emphasizes how complex our "target" encoding has become, but didn't cause this complexity. The way I encoded columns with weird names in this patch is identical to what Cassandra does (basically, the CQL quoting, our `util::maybe_quote()`), I don't know if we care about compatibility with Cassandra in this area, but I guess it can't hurt.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1000/

nyh commented:
Ping.

avikivity commented:
I'll review it.

avikivity commented on cql3/Cql.g:
```diff
@@ -849,7 +849,8 @@ indexIdent returns [::shared_ptr<index_target::raw> id]
     @init {
         std::vector<::shared_ptr<cql3::column_identifier::raw>> columns;
     }
-    : c=cident                   { $id = index_target::raw::values_of(c); }
```
This first patch doesn't fix #8745, the entire series does. The first patch alone isn't sufficient.

nyh replied:
Issue #8745 was not about fully implementing VALUES() but just adding its syntax. Adding its syntax means we get the right error message for values(v) when v is a **frozen** set (the error should be that v is frozen, not a syntax error). I think it is fine that we mark 8745 fixed as soon as the syntax is added - even if values(v) for a non-frozen set still doesn't work (just like keys(v), etc., doesn't work).

avikivity commented on index/secondary_index_manager.cc:
```diff
@@ -96,9 +96,6 @@ view_ptr secondary_index_manager::create_view_for_index(const index_metadata& im
     auto target_type = target_info.type;
-    if (target_type != cql3::statements::index_target::target_type::values) {
-        throw std::runtime_error(format("Unsupported index target type: {}", to_sstring(target_type)));
-    }
```
What happens if some creates such an index with only this patch?

avikivity commented on column_computation.hh:
```diff
@@ -36,7 +36,7 @@ public:
 
     virtual bytes serialize() const = 0;
-    virtual bytes_opt compute_value(const schema& schema, const partition_key& key, const clustering_row& row) const = 0;
+    virtual bytes compute_value(const schema& schema, const partition_key& key) const = 0;
```
I mentioned before that general computed column can and do depend on the row, and can return null (for example, if an input they depend on is null). I suppose we can undo this when we implemented general computed columns.

nyh replied:
My first comment on this patch was also that I don't understand why do this change it all. If the "row" parameter was unused, we could have kept it unused and not fix it today. 

But I agree with @enedil that if the new code does not use the row key (it does use the row and get it from some other mechanism!) and doesn't return null ever (unless I'm missing something?) then new API of this function is fine, and we don't to anticipate the future - we can always change it again later when the need arises.

We just need to make a decision here.

psarna replied:
I'm fine with leaving the original `bytes_opt` return value, it makes this already big series a little bit smaller.

nyh replied:
I don't know what to do about this. Both ways (leave it bytes_opt as it was in the past, or leave it bytes as it is now in this patch) make some sense, but leaving it as it is **now** (bytes) is a bit easier for me because I don't need to change anything...
Another reason why I think it makes sense to leave it as it is now (bytes), is that the old code checked if the opt_bytes was null, and if it was, did on_internal_error. So maybe it's better just not to have this option?

avikivity commented on schema.cc:
```diff
@@ -1704,9 +1704,9 @@ bytes token_column_computation::serialize() const {
+bytes token_column_computation::compute_value(const schema& schema, const partition_key& key) const {
     auto long_value = dht::token::to_int64(dht::get_token(schema, key));
-    return long_type->decompose(long_value);
+    return {long_type->decompose(long_value)};
```
Why the new braces?

nyh replied:
They are not needed. I removed them now.

avikivity commented on column_computation.hh:
```diff
@@ -13,6 +13,12 @@
+
+namespace db::view {
+struct bytes_with_action;
+}
```
In the previous incarnation of this series we had a long discussion about why I didn't like bytes_with_action. What happened to it?

nyh replied:
Nothing. I took @enedil 's patches as-is (just fixing a few small conflicts and problems) and added specific patches at the end. If you didn't like bytes_with_action, your comments probably still apply and I can take a look at them.

Can you please either copy the relevant copy from that PR to here, or, alternatively, go to the old PR and "resolve" your comments which were either addressed or you don't insist on any more? Because I understood from @psarna that for some of the things you wanted to redesign, the decision was not to do that redesign. If there are things that still need to be redesigned before committing this series, I want to understand what those are. Thanks.

avikivity replied:
I forgot the details, maybe @psarna remembers. If not I can try to recreate them.

nyh replied:
The original review was https://github.com/scylladb/scylla/pull/10123 (as I noted, it is not possible to change the ownership of a PR, which is why I was forced to create this new one). Maybe you can take a look at that and just "resolve" your comments that you don't care about any more, or were handled, and then I'll go there and also address your old comments that still need addressing.

nyh replied:
I see your comments about bytes_with_action in https://github.com/scylladb/scylla/pull/10123/commits/4389d2e107db8c80b6930898c7097c385022d167 but I didn't understand what exactly your complaint is :-( Since I also found this class hard to understand, I can try to improve it the way I see fit - but I can't be sure that I'll improve it in the same way that you want to improve it, so it will be good if you can tell me more about what exactly you didn't like.

avikivity replied:
A computed column is a pure computation where the computed column's value is derived from some other columns. But that's not the case for bytes_with_actions - each individual "action" gets distributed into a new row. So it's abusing the column computation abstraction.

I don't remember if bytes_with_action violates the pureness property or not.

psarna replied:
My conclusion about the bytes_with_action discussion was posted here: https://github.com/scylladb/scylla/pull/10123#issuecomment-1160101326

tl;dr: one of the discussed solutions for this was to rename "column_computation" to something else, now that it starts serving a slightly different purpose, producing multiple rows. Technically I still think that the name "column computation" doesn't really imply a single value, because a column is a subset of values for given table, but that's bikeshedding, so I'm also absolutely fine with figuring out a different, more descriptive name. I'm not convinced by the alternative suggested here: https://github.com/scylladb/scylla/pull/10123#issuecomment-1157529922, for more context I recommend reading my original comment above.

avikivity replied:
Column computation doesn't imply a single value, but "computed column" implies storing that value (perhaps of collection type) once, not generating multiple rows.

If we're renaming, let's also rename bytes_with_actions. It's incredibly non-descriptive. What are those bytes? and what are the actions?

nyh replied:
I still don't understand how renaming or refactoring or re-explaining **bytes_with_action** is related to renaming "column_computation"  or changing its return value from a vector of whatever, to something else.

I'll focus my efforts now on renaming or cleaning up bytes_with_actions itself - let me know if you want me to do something else.

nyh replied:
The more I look at this collection_column_computation code, the less I think it should be generic. The base class column_computation continues to support only a single return value - not a list. The subclass collection_column_computation has a new method returning a list - and it is only called in one place, in the view.cc code. That place actually use dynamic_cast to give collection_column_computation very special treatment - that no other future column_computation can get if it also wants to return multiple results.

I'm really tempted moving all this collection_column_computation code into a separate source file that lives in db/view/ and stop pretending it is something general. 

But before I do that, @avikivity @psarna I want to understand if one of you would object, and actually wanted the feature of computed column returning multiple values to be a generic feature that later we can reuse without change. If the answer is yes, I think the way it is written now is not general enough.

psarna replied:
Pardon me, I somehow missed this comment 2 days ago - I also agree that making things generic without a very good reason is a mistake, so moving this code to a separate file with specialized implementation sounds good. And as for naming bytes_with_action, I don't have any good suggestions, but here's a mediocre one: `value_or_tombstone_or_marker`

nyh replied:
Sadly, the more I look at it, the less sense bytes_with_actions makes. get_view_rows() actually returns not just bytes_with_action, but a row *and* an action. So it views bytes_with_action as an additional action, not the main row addition, but in the collection column indexing, there is *only* this bytes_with_action.

I'm trying now to write comments around this, but this organization is very hard to explain or defend. On the other hand, I don't see what other alternative I have besides spending many days rewriting this already-working code.

avikivity commented on db/view/view.cc:
```diff
@@ -478,41 +480,187 @@ row_marker view_updates::compute_row_marker(const clustering_row& base_row) cons
+struct value_getter {
+    // linearized_values hold bytes for values of computed columns, for which we later store references in the form of managed_bytes_view.
+    // deque doesn't invalidate references at emplace_back.
+    std::deque<bytes> linearized_values;
```
What happens if a collection value causes a clustering key that is too large to be generated? Or is that impossible?

nyh replied:
This is a good question. If I remember correctly, while old Cassandra had a 64KB limit on collection values, this limit was lifted, while clustering keys do have a 64KB limit.
I'll write a test for this, and fix it if it doesn't have reasonable behavior.

nyh replied:
https://github.com/scylladb/scylla/issues/10948

nyh replied:
That was a duplicate. The original issue: https://github.com/scylladb/scylla/issues/8627
I'm working now on the tests, and the fix.

nyh replied:
To clarify, the current code - in both regular indexing and the new collection indexing - can indeed generate an oversized clustering key. The write of this clustering key will fail. But the way it fails looks like an obscure internal server error instead of a clean InvalidRequest saying that the query is broken and why.
It's not a disaster (Scylla doesn't crash...) but it still needs to be fixed, and I'm planning to add another patch to fix it (both in the collection indexing case and in the regular indexing case).

nyh replied:
@eliransin @psarna FYI I'm working on fixing this comment from Avi (which is https://github.com/scylladb/scylla/issues/8627) right now. It's taking me longer than I hoped because I'm rusty with the materialized views code, which has also become so complex that it's really hard to get anything done there :-(

nyh replied:
I decided that fixing https://github.com/scylladb/scylla/issues/8627 is harder than it's worth, so will not do it now. Instead I will add to this series as a 28th patch more test cases for this issue (oversized items in a collection that can't be indexed). The test will xfail, but will demonstrate that the oversized keys still fail in an "acceptable" manner (replica write error) and not cause any crashes or other disasters.
@avikivity do you think that's acceptable?

avikivity replied:
yes

avikivity commented on db/view/view.cc:
```diff
@@ -478,41 +480,187 @@ row_marker view_updates::compute_row_marker(const clustering_row& base_row) cons
+    const partition_key& base_key;
+    const clustering_row& update;
+    const std::optional<clustering_row>& existing;
+
```
Why are these three members not following the coding style?

nyh replied:
Done. I pushed now a new version with added underscores on these variables.

avikivity commented on cql3/statements/select_statement.cc:
```diff
@@ -1341,7 +1344,24 @@ indexed_table_select_statement::find_index_clustering_rows(query_processor& qp,
+                    }
+                }
+            }
+
```
What happens when paging, if a page boundary splits this check?

nyh replied:
Good question. I'll write a test for this. If it doesn't work (and I'm worried it doesn't), I don't know how to fix it. But I'll cross that bridge when I come to it (after writing a test).

nyh replied:
@avikivity was absolutely right, this is a bug, and I confirmed it exists in a test. I'll push the xfailing test to the ever-growing series, but then need to think whether to solve this now, and how.

nyh replied:
Pushed patch 32 with a test for this bug.

nyh replied:
Pushed patch 33 ("secondary index: fix paging in map value indexing") fixing this bug and making the tests for it pass.

nyh replied:
I made the test stronger and found a variant of this bug still exists in the special case where the base table has only partition keys. So need to fix that too :-( What a mess :-(

nyh replied:
I know where to fix the case of only partition keys (there's separate code for that case), but it's harder to fix. This whole indexing code is so complicated and so interconnected :-(

nyh replied:
I have good news and bad news. The good news is that I actually fixed the case for partition keys correctly.

The bad news is that in both partition-key-only and partition-and-clustering-key cases, if there are any **empty pages** in the middle of the response - because a page of one item was a duplicate and discarded - the query stops prematurely. Only my partition-key test had this case which is why I thought the new bug was specific to partition keys.

nyh replied:
It turns out that I fell in exactly the **driver bug** about which I told @denesb about today - older versions of the Cassandra/Scylla drivers stop reading when reaching an empty page, which is a bug and was fixed later. I have a fixture `driver_bug_1` which skips the test if the driver is too old, and now I added it and the failing tests got skipped. 

Now I need to update the driver to see if these tests now pass.

nyh replied:
Ok, good news: I pushed a new patch 33 which makes the tests stronger and passes all the tests. I had to fix the partition-key-only and partition-and-clustering-key code in two separate places, but using exactly the same technique and it's fairly straightforward. So I think I'm done with this TODO as well, and perhaps this series is finally ready for prime time?

avikivity commented:
Missing cluster feature to prevent creating such indexes in a partially-upgraded cluster.

nyh commented:
Rebased, did one small review comment, and added a 28th patch with extensive tests answering @avikivity 's review question: What happens when an indexed value is larger than a key allows (64K). The tests demonstrates several angles of this problem (index, view, view of collection, regular update vs. view building of existing data), on which we already have an issue https://github.com/scylladb/scylla/issues/8627.

This issue should eventually be solved, but it's not easy and I decided not to do it now. Importantly, as the tests in the 28th show, this problem is not new or specific to collection indexing, plus now we have a specific test for collection indexing so when we are ready to solve that issue, we'll also solve it for the collection-indexing case.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1159/

AugustDev commented:
This is incredibly useful feature, any ETA on merging it?

nyh commented:
> This is incredibly useful feature

I agree. The issue had 20 people "voting" on it :-)

> any ETA on merging it?

The feature is basically ready and this PR applies to the master branch (it hasn't bitrotted), so if you just want to try it out, you can do it immediately, now.

Regarding when it can be officially merged, this depends on the reviewers. This code adds A LOT of code, and there are still some places which the reviewers (rightly) regard as too obscure or complex, and I will need to rewrite.

nyh commented:
Unfortunately, this series bitrotted a bit, so I'm now trying to fix it and will push a new version.

We have in cql3/expr/to_restriction.cc code which says that an expression cannot operate on a collection. In particular the mandatory "x IS NOT NULL" of materialized views, when used on a collection column x, generates an error.

The strange thing is that this error check is already two months old, but apparently that code path wasn't used for materialized views, and now it is. If I remember correctly, the "IS NOT NULL" specification was never treated as a real expression and was never evaluated (IIRC, the MV code just skipped those rows) but apparently it is now a real expression and really evaluated, and we also need to allow IS NOT NULL on a collection.

CC @cvybhu

nyh commented:
I have this trivial patch
```patch
diff --git a/cql3/expr/restrictions.cc b/cql3/expr/restrictions.cc
index 5cad25582..534e1504a 100644
--- a/cql3/expr/restrictions.cc
+++ b/cql3/expr/restrictions.cc
@@ -21,7 +21,7 @@ extern logging::logger expr_logger;
 namespace {
 
 bool is_legal_relation_for_non_frozen_collection(oper_t oper, bool is_lhs_col_indexed) {
-    return oper == oper_t::CONTAINS_KEY || oper == oper_t::CONTAINS || (oper == oper_t::EQ && is_lhs_col_indexed);
+    return oper == oper_t::CONTAINS_KEY || oper == oper_t::CONTAINS || oper == oper_t::IS_NOT || (oper == oper_t::EQ && is_lhs_col_indexed);
 }
```
and it fixes the problems, but I'm not happy. The expression "m IS NOT NULL" for a map, while may make some sense, is not really relevant to the case where we index the **keys** or **values** of m, not m itself. An alternative is to drop the silly "m IS NOT NULL" from the secondary index's MV definition. I'll try this approach as well.

nyh commented:
Pushed a new version - it isrebased to fix what bitrotted, and also added yet another new patch (18/29) https://github.com/scylladb/scylla/pull/10841/commits/c519cd7c74d948d6b7103e13270e5986068b3f3e which fixes a new problem that appeared due to changes in the expression code:

When secondary-index code creates a materialized view for indexed column x, it uses "x IS NOT NULL" in the definition of the view, as required. It also did this if we don't really index x but a computed column (i.e., x is a collection column). But after recent changes to the expression code, "x IS NOT NULL" for a collection column started failing. The solution is not to add the "x IS NOT NULL" when indexing a collection.

Hopefully with those fixes, the series will start passing all the tests again.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1396/

nyh commented:
> Do we still have bytes_with_action here or has it morphed into something else?

Yes, I don't know what to do about it. I'll about this later.

nyh commented:
> Missing cluster feature to prevent creating such indexes in a partially-upgraded cluster.

Good catch. Added a 30th commit in the series to do that (it's very easy nowadays, kudos on simplifying the cluster-feature setup!)

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1426/

nyh commented:
Pushed a new version, rebased to fix conflict and renamed and commented the "bytes_with_action" structure.

The new name of this structure is "view_key_and_action", which suggests that each such object (we have multiple when indexinga  collection) is about one view key (where to add a view row), and an additional "action" that we need to take beyond adding the view row. The comment in front of this struct also explains this in detail.

This is the best I can do to make this code easier to understand without completely reorganizing it.
@avikivity @psarna please take a look (it's just the last patch) and tell me if you think it's good enough.

There are also a couple of other review comments I still need to address, but I think that's it.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1496/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1504/

psarna commented:
Oh, that sucks. Looks like we have at least three options here:
 1. accept this bug as non-critical and temporarily document it until it gets fixed
 2. create an ugly special case for collection indexing, e.g. to continue reading no matter what until we read a non-duplicate row
 3. add a special case in our paging state, adding custom state to it specifically for the purpose of deduplicating collection indexing

Any other suggestions for a solution/hack/workaround? All of these sound bad, but we need to pick something. Option (1.) has one advantage of any other option - no code changes, and it's already a fragile area.

nyh commented:
> Oh, that sucks. Looks like we have at least three options here:
> 
>     1. accept this bug as non-critical and temporarily document it until it gets fixed
> 
>     2. create an ugly special case for collection indexing, e.g. to continue reading no matter what until we read a non-duplicate row
> 
>     3. add a special case in our paging state, adding custom state to it specifically for the purpose of deduplicating collection indexing
> 
> 
> Any other suggestions for a solution/hack/workaround? All of these sound bad, but we need to pick something. Option (1.) has one advantage of any other option - no code changes, and it's already a fragile area.

Yes, I already have a test for option 1 - I would just need to create a new issue :-)
I was hoping maybe I can detect a duplicate on a followup page by checking if I find another row which is identical (in base-table key) to the beginning of this page. I think this will work, but I'll need to see if/how I can code it.

By the way, I'm going on vacation in a couple of days and I don't know what will happen if we don't get it in by then :-( This turned out to be a much bigger problem than I thought it would be when I took over it...

nyh commented:
> I was hoping maybe I can detect a duplicate on a followup page by checking if I find another row which is identical (in base-table key) to the beginning of this page. I think this will work, but I'll need to see if/how I can code it.

Luckily, this approach worked! I still need to write and run more tests that this really works in various cases, but it did pass the tests I already wrote. All I needed to do is when paging, set last_primary_key to the base key (without the extra components we add in the view) in the paging state and not start it empty:
```c++
        // Set last_primary_key if indexing collection values and not in the
        // first query page. See comment below why last_primary_key is needed
        // in that case. We have a test for this case with paging:
        // test test_secondary_index.py::test_index_map_values_paging.
        std::optional<primary_key> page_start_primary_key;
        if (_index.target_type() == cql3::statements::index_target::target_type::collection_values &&
            options.get_paging_state()) {                
            auto paging_state = options.get_paging_state();
            auto base_pk = generate_base_key_from_index_pk<partition_key>(paging_state->get_partition_key(),
                paging_state->get_clustering_key(), *_schema, *_view_schema);
            auto base_dk = dht::decorate_key(*_schema, base_pk);
            auto base_ck = generate_base_key_from_index_pk<clustering_key>(paging_state->get_partition_key(),
                paging_state->get_clustering_key(), *_schema, *_view_schema);
            page_start_primary_key = primary_key{std::move(base_dk), std::move(base_ck)};
            last_primary_key = *page_start_primary_key;
        }
```

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1523/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1526/

nyh commented:
@avikivity @psarna and others - I think I have addressed all the review comments.

I can't say I'm in love with the new code - it took an already complicated and difficult-to-follow piece of code written piece by piece over years - and made it even more complicated and more difficult to follow. I tried to improve whatever I could **without** completely rewriting this series - adding comments and renaming that bytes_with_actions thing. As you saw I also wrote more tests and fixed several bugs (and also wrote more tests that confirmed that some things were already correct).

I don't know how to make this series even more ready for commit without spending another couple of weeks completely rewriting it, which I'm not thrilled to do. I'm going on summer vacation at the end of this week, by the way.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1542/

nyh commented:
Rebased to most recent master (not other changes, I believe I'm done with this series and waiting for reviewers)

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1771/

psarna commented:
I'm extremely hesitant about this code, but it looks like it doesn't affect the regular path (non-collection indexing) in a bad way, so it's better to push it rather than let it rot. But meanwhile, it already got a conflict in view.cc, so @nyh let me know once you rebase and we can proceed with moving this series forward once CI passes. And then, we can start rebasing #11166 on top of it...

nyh commented:
> I'm extremely hesitant about this code, but it looks like it doesn't affect the regular path (non-collection indexing) in a bad way, so it's better to push it rather than let it rot. But meanwhile, it already got a conflict in view.cc, so @nyh let me know once you rebase and we can proceed with moving this series forward once CI passes. And then, we can start rebasing #11166 on top of it...

I rebased to latest master to fix the conflict.

I feel the same way as you - although I reviewed this code, my impression is that in some places it continues to complicate code which was **already** too hard for me to fully understand - so I don't feel 100% confident that I was able to understand everything and be sure that the changes do not introduce new bugs in pre-existing code. 

However, I think there's a limit to what we can ever hope to achieve with reviews (self-review of the author, and additional reviews by reviewers), and to trust refactoring patches (like this one) we need to count on good test coverage. I think we do have fairly good test coverage in the materialized views area (Duarte wrote a lot of tests in C++), but I don't believe it is completely exhaustive, unfortunately. Perhaps we can do something using @denesb 's framework for testing test coverage, here checking the test coverage of materialized views tests?

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1806/

avikivity commented:
So what happens with bytes_in_actions? Are we committing this? It's a crime against future readers of the code.

What about the persistent state? Does it also contains things like bytes_with_actions?

nyh commented:
I added a patch near the end titled "rename and explain bytes_with_action" which 1. Renames it to something more sensible, 2. Explains better (I hope) what it means.

avikivity commented:
> I added a patch near the end titled "rename and explain bytes_with_action" which 1. Renames it to something more sensible, 2. Explains better (I hope) what it means.

Ok. I can't say I like it but it's certainly better.

What about the persistent footprint in schema_tables? How does it look?

nyh commented:
> > I added a patch near the end titled "rename and explain bytes_with_action" which 1. Renames it to something more sensible, 2. Explains better (I hope) what it means.
> 
> Ok. I can't say I like it but it's certainly better.

I agree on both counts ;-)

> What about the persistent footprint in schema_tables? How does it look?

I assume you mean the "index target" issue?
I fixed (and explained) this in the commit titled "cql: fix secondary index "target" when column name has special characters".
Here is a quote of that commit's message, with some minimal edits:

We encode the "target" of a secondary index in one of three ways:
1. It can be just a column name
2. It can be a string like keys(colname) - for the new type of collection indexes introduced in this series.
3. It can be a JSON map ({ ... }). This form was used already for local indexes.

The code parsing this target - target_parser::parse() - needs not to confuse these different formats. Before the aforementioned patch, if the column name contains special characters like braces or parentheses (this is allowed in CQL syntax, via quoting), we can confuse case 1, 2, and 3: A column named "keys(colname)" will be confused for case 2, and a column named "{123}" will be confused with case 3.

The solution adopted in this patch is that the column name in case 1 should be escaped somehow so it cannot be possibly confused with either cases 2 and 3. The way we chose is to convert the column name to CQL (with  `column_definition::as_cql_name()`). In other words, if the column name contains non-alphanumeric characters, it is wrapped in quotes and also quotes are doubled, as in CQL. The result of this can't be confused with case 2 or 3, neither of which may begin with a quote.
    
This escaping is not the minimal we could have done, but incidentally it is exactly what Cassandra does as well, so I used it as well.    

This change is *mostly* backward compatible: Already-existing indexes will  still have unescaped column names stored for their "target" string, and the unescaping code will see they are not wrapped in quotes, and not change them. Backward compatibility will only fail on existing indexes on columns whose name begin and end in the quote characters - but this case is extremely unlikely.

This patch illustrates how un-ideal our index "target" encoding is, but isn't what made it un-ideal. We should not have used three different  formats for the index target - the third representation (JSON) should have sufficed. However, two two other representations are identical to Cassandra's, so using them when we can has its compatibility advantages.

psarna commented:
Alright, it looks like the second run of this series (originally: #10123) already added test coverage, fixed a big bunch of issues, sprinkled comments and more descriptive names all over the place, is guarded by a schema feature and is unlikely to break existing code paths (esp. for non-collection indexing), so I'll close my eyes and queue it to next. Cross your fingers

avikivity commented:
> > > I added a patch near the end titled "rename and explain bytes_with_action" which 1. Renames it to something more sensible, 2. Explains better (I hope) what it means.
> > 
> > 
> > Ok. I can't say I like it but it's certainly better.
> 
> I agree on both counts ;-)
> 
> > What about the persistent footprint in schema_tables? How does it look?
> 
> I assume you mean the "index target" issue? I fixed (and explained) this in the commit titled "cql: fix secondary index "target" when column name has special characters". Here is a quote of that commit's message, with some minimal edits:
> 
> We encode the "target" of a secondary index in one of three ways:


I meant how does bytes_with_actions appear in the materialized view schema_tables.

psarna commented:
> > > > I added a patch near the end titled "rename and explain bytes_with_action" which 1. Renames it to something more sensible, 2. Explains better (I hope) what it means.
> > > 
> > > 
> > > Ok. I can't say I like it but it's certainly better.
> > 
> > 
> > I agree on both counts ;-)
> > > What about the persistent footprint in schema_tables? How does it look?
> > 
> > 
> > I assume you mean the "index target" issue? I fixed (and explained) this in the commit titled "cql: fix secondary index "target" when column name has special characters". Here is a quote of that commit's message, with some minimal edits:
> > We encode the "target" of a secondary index in one of three ways:
> 
> I meant how does bytes_with_actions appear in the materialized view schema_tables.

From the point of view of schema tables, it's just yet another computed column, here's its serialization code, it's a json: https://github.com/scylladb/scylladb/pull/10841/commits/ebc4ad471321697cb58094fbf61a778a3e05cece (grep for `serialize(`)

nyh commented:
> > > > > I added a patch near the end titled "rename and explain bytes_with_action" which 1. Renames it to something more sensible, 2. Explains better (I hope) what it means.
> > > > 
> > > > 
> > > > Ok. I can't say I like it but it's certainly better.
> > > 
> > > 
> > > I agree on both counts ;-)
> > > > What about the persistent footprint in schema_tables? How does it look?
> > > 
> > > 
> > > I assume you mean the "index target" issue? I fixed (and explained) this in the commit titled "cql: fix secondary index "target" when column name has special characters". Here is a quote of that commit's message, with some minimal edits:
> > > We encode the "target" of a secondary index in one of three ways:
> > 
> > 
> > I meant how does bytes_with_actions appear in the materialized view schema_tables.
> 
> From the point of view of schema tables, it's just yet another computed column, here's its serialization code, it's a json: [ebc4ad4](https://github.com/scylladb/scylladb/commit/ebc4ad471321697cb58094fbf61a778a3e05cece) (grep for `serialize(`)

That's not "bytes_with_action" (now - view_key_and_action) - what you (Piotr) are describing here is the "computed column". The bytes_with_action is not serialized, it was just a hard to understand structure to pass parameters between internal functions.

avikivity commented:
I meant did the schema thing that describes that we want to compute bytes_with_actions rather than the token.

QuickListTech commented:
Which version of scylla is this feature going into? i'm excited!

nyh commented:
> Which version of scylla is this feature going into? i'm excited!

It missed open-source version 5.1 or enterprise version 2022.1, as those are already being tested, so it should be in the following version. You can already test it now if you compile Scylla yourself, or use a nightly image (of course, you shouldn't do that in production).

tzach commented:
@nyh FYI
the first example in the PR
`CREATE TABLE test(int id, somemap map<int, int>, somelist<int>, someset<int>, PRIMARY KEY(id));`
should be 
`CREATE TABLE test(id int, somemap map<int,int>, somelist list<int>, someset set<int>, PRIMARY KEY (id));`
